src/common/com/intellij/plugins/haxe/ide/refactoring/introduce/HaxeIntroduceHandler.java

```diff
@@ -319,15 +324,23 @@ private static boolean notFunctionMethodClass(HaxeComponentType type) {
     if (expression instanceof HaxeCallExpression) {
       final HaxeExpression callee = ((HaxeCallExpression)expression).getExpression();
       text = callee.getText();
+    } else if (expression instanceof HaxeRegularExpression) {
+      text = "REGEX";
+    } else if (expression instanceof HaxeLiteralExpression) {
+      text = "CONST_" + expression.getText().toUpperCase();
     }
 
     if (text != null) {
       candidates.addAll(HaxeNameSuggesterUtil.generateNames(text));
     }
+    if (candidates.isEmpty()) {
+      candidates.add("x");
+    }
```

Comment:
Always need at least one, or we get NPEs down in the CodeInsight layers.

Reply:
That's a good thought, and that's really what happens:

The algorithm that calculates names basically removes all non-letter characters, then concatenates all letter character groups (words) with an underscore between them.  Then it tries various versions of the string getting shorter (by removing leading word groups) and creates a list of them shortest first.  For example `var a : Int = foo + bar` turns into `var_a_Int_foo_bar` and then is parsed into a list like so: `var_a_Int_foo_bar`, `int_foo_bar`, `foo_bar`, `bar`, which is then reversed and returned.  (And, maybe that should be improved.)

Once the list is generated, each candidate is compared to existing variables (and now keywords) and, if duplicated, gets added a sequential number (e.g. `bar1`).  The shortest is automatically selected and used for display, the others are put into a dropdown box for selection (like completion).

This single letter is a last-ditch solution for when nothing else occurs.  This doesn't happen often, but will when the expression is a set of numeric literals with operators in between.  Then, I guess, I could go the extra mile here to figure out the type and use a better letter.  But that case will *always* be a float, so I could just go with 'f', I guess.

Reply:
@mayakwd  So, I thought to myself, let's go look at what Java does... and you're right, they do more.  So, I figured I could just hook their nameUtil and it would be done.  But, no.... that was just the start.  :smiley:  It took me a while and I found some cases where it wasn't working.  And, now, another week's worth of work later (and having to implement resolution for type checks), I ready for a re-review.